### PR TITLE
Carving out -Wformat warning about scoped enums into a subwarning

### LINKED ISSUE #81647: Carving out -Wformat warning about scoped enums into a subwarning
Android uses -Wformat by default.  Starting with https://github.com/llvm/llvm-project/commit/3632e2f517, this warning reports a lot of conversion warnings for scoped enums.  It is infeasible to fix these in our third party code.

We can create a sub-warning, similar to other named diagnostics controlled by -Wformat (https://clang.llvm.org/docs/DiagnosticsReference.html#wformat), to be able to disable it selectivey for our third-party code.

@AaronBallman does this option sound good?

(cc: @enh-google @ZijunZhaoCCK )

### PR DESCRIPTION
**Pull request by ZijunZhaoCCK**
Make it part of -Wformat-pedantic.

Fixes #81647

### CODE REVIEW

File: clang/test/Sema/format-strings-signedness.c [hunk 1]
```diff
@@ -12,8 +12,8 @@
 // RUN: %clang_cc1 -triple=x86_64-pc-linux-gnu -std=c11 -fsyntax-only -Wformat -verify=okay %s
 
 // Verify that -Wformat-signedness with -Wno-format are not reported (gcc compat).
```
[the remark this answers is not in the document]
Comment (AaronBallman) [reply clang/test/Sema/format-strings-signedness.c:14]: Yes, this is intentional, see the discussion at: https://github.com/llvm/llvm-project/issues/81647#issuecomment-2030070508

**Review by AaronBallman — COMMENTED** [review]
Please mention that this fixes #81647 in the patch summary so that the issue is automatically closed.

You should also add a release note so that users know about the change (and link to the issue being closed).

File: clang/test/Sema/format-strings-signedness.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (karka228) [reply clang/test/Sema/format-strings-signedness.c]: That discussion is only about -Wformat scoped enums warnings, but this patch currently move all the -Wformat warnings regarding type mismatch into a subwarning, that a lot more warnings than only scoped enums. Is that intended?

[the remark this answers is not in the document]
Comment (AaronBallman) [reply clang/test/Sema/format-strings-signedness.c]: Agreed, that seems like an outstanding issue with the PR and should be addressed. Only the scoped enumeration behavior should change.

File: clang/test/FixIt/format.cpp [hunk 2]
```diff
@@ -1,5 +1,6 @@
 // RUN: %clang_cc1 -fsyntax-only -verify -Wformat %s
-// RUN: %clang_cc1 -fsyntax-only -fdiagnostics-parseable-fixits -Wformat %s 2>&1 | FileCheck %s
+// RUN: %clang_cc1 -fsyntax-only -fdiagnostics-parseable-fixits -Wformat-pedantic %s 2>&1 | FileCheck %s
```
Comment (pirama-arumuga-nainar) [inline clang/test/FixIt/format.cpp:2]: With the RUN: line including `-Wformat-pedantic`, do we need the rest of the changes (adding `expected-no-diagnostics` and the removal of `expected-warning`s)?

**Comment by pirama-arumuga-nainar** [comment]
> Please mention that this fixes #81647 in the patch summary so that the issue is automatically closed.
> 
> You should also add a release note so that users know about the change (and link to the issue being closed).

This is also pending.  The rest of the changes LGTM.

File: clang/test/FixIt/format.cpp [hunk 3]
```diff
@@ -1,5 +1,5 @@
-// RUN: %clang_cc1 -fsyntax-only -verify -Wformat %s
-// RUN: %clang_cc1 -fsyntax-only -fdiagnostics-parseable-fixits -Wformat %s 2>&1 | FileCheck %s
+// RUN: %clang_cc1 -fsyntax-only -verify -Wformat-pedantic %s
```
Comment (AaronBallman) [inline clang/test/FixIt/format.cpp:1]: We're missing a comparison between `-Wformat` and `-Wformat-pedantic` in terms of handling scoped enumerations. We should have a pair of RUN lines to show that this diagnostic is in pedantic only.

One thing that helps with testing this is to use something like `-verify=expected,pedantic` and `// pedantic-warning {{msg}}`: https://clang.llvm.org/docs/InternalsManual.html#verifying-diagnostics

**Review by AaronBallman — COMMENTED** [review]
Please be sure to add a release note in clang/docs/ReleaseNotes.rst. The code changes themselves LGTM, but there are some issues with the tests.

**Review by AaronBallman — APPROVED** [review]
LGTM!

**Pull request merged**